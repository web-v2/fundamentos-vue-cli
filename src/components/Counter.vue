<template>
<div>
    <h2>{{customTitle}}</h2>
    <p>{{count}}<sup>2</sup> = {{squareCounter}}</p> 
    <p data-test-id="counter">{{count}}</p>
    <div class="container-vue">        
        <button @click="increaseCounter">+1</button>
        <button @click="decreaseCounter">-1</button>
    </div>
</div>
</template>

<script>

export default {
    props: {
        title: String,
        start: {
            type: Number,
            default: 100,
            validator(value){
                return value > 0
            }
        }
    },
    data() {
        return {
            count: this.start
        }
    },
    methods: {
        getSquareCount(){
            return this.count * this.count
        },
        increaseCounter(){
            return this.count++;
        },
        decreaseCounter(){
            return this.count--;
        }
    },
    computed: {
        squareCounter(){
            return this.count * this.count
        },
        customTitle(){
            return this.title || 'Counter'
        }
    }
}
</script>

<style>
    button{
        background-color: #64B687;
        border-radius: 5px;
        border: 1px solid #fff;
        color: #fff;
        cursor: pointer;
        margin: 0 5px;
        padding: 5px 15px;
        transition: 0.3s ease-in-out;
    }
    button:hover{
        background-color: #5aa67b;
        transition: 0.3s ease-in-out;
    }
</style>