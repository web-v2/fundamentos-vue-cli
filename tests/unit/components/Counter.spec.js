import { shallowMount } from '@vue/test-utils';
import Counter from '@/components/Counter.vue';

describe('Counter Componente', () => {
    let wrapper;
    beforeEach(() => {
        wrapper = shallowMount(Counter);
    });
    test('Debe de hacer match con el snapshot', () => {        
        expect( wrapper.html() ).toMatchSnapshot();
    });
    test('H2 debe tener valor por defecto "Counter"', () => {        
        expect( wrapper.find('h2').exists() ).toBe(true);
        const h2Value = wrapper.find('h2').text();
        //console.log(h2.text());
        expect( h2Value ).toBe('Counter');
    });
    test('El valor por defecto debe ser 100 en p', () => {        
        const pTag = wrapper.find('[data-test-id="counter"]').text();
        expect( pTag ).toBe('100');
    });
    test('Debe de incrementar y decrementar el contador ', async() => {

        const [increaseBtn, decreaseBtn] = wrapper.findAll('button');
        await increaseBtn.trigger('click');
        await increaseBtn.trigger('click');
        await increaseBtn.trigger('click');                
        await decreaseBtn.trigger('click');
        await decreaseBtn.trigger('click');

        const pTag = wrapper.find('[data-test-id="counter"]').text();
        expect( pTag ).toBe('101');
    });    
    test('Debe de establecer el valor por defecto ', () => {
        const {start} = wrapper.props();
        const valor = wrapper.find('[data-test-id="counter"]').text();
        expect( Number(valor) ).toBe(start);
    });
    test('Debe de mostrar la prop title', () => {
        const title = 'Hello World';
        const wrapper = shallowMount(Counter, {
            props: {
                title
            }
        });
        expect( wrapper.find('h2').text() ).toBe(title);
    });
});