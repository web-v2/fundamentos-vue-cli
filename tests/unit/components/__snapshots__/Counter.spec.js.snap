// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Componente Debe de hacer match con el snapshot 1`] = `
<div>
  <h2>Counter</h2>
  <p>100<sup>2</sup> = 10000</p>
  <p data-test-id="counter">100</p>
  <div class="container-vue"><button>+1</button><button>-1</button></div>
</div>
`;
